sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,492 @@
+      abstract Read build();
+    }
+
+    public Read withMqttConnectionConfiguration(MqttConnectionConfiguration configuration) {
+      checkNotNull(configuration, "MqttConnectionConfiguration");
+      return toBuilder().setMqttConnectionConfiguration(configuration).build();
+    }
+
+    public Read withMaxNumRecords(long maxNumRecords) {
+      return toBuilder().setMaxNumRecords(maxNumRecords).build();
+    }
+
+    public Read withMaxReadTime(Duration maxReadTime) {
+      return toBuilder().setMaxReadTime(maxReadTime).build();
+    }
+
+    @Override
+    public PCollection<byte[]> apply(PBegin input) {
+
+      org.apache.beam.sdk.io.Read.Unbounded<byte[]> unbounded =
+          org.apache.beam.sdk.io.Read.from(new UnboundedMqttSource(this));
+
+      PTransform<PBegin, PCollection<byte[]>> transform = unbounded;
+
+      if (maxNumRecords() != Long.MAX_VALUE) {
+        transform = unbounded.withMaxNumRecords(maxNumRecords());
+      } else if (maxReadTime() != null) {
+        transform = unbounded.withMaxReadTime(maxReadTime());
+      }
+
+      return input.getPipeline().apply(transform);
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      mqttConnectionConfiguration().validate();
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      mqttConnectionConfiguration().populateDisplayData(builder);
+      builder.add(DisplayData.item("maxNumRecords", maxNumRecords()));
+      builder.addIfNotNull(DisplayData.item("maxReadTime", maxReadTime()));
+    }
+
+  }
+
+  private static class UnboundedMqttSource
+      extends UnboundedSource<byte[], UnboundedSource.CheckpointMark> {
+
+    private Read spec;
+
+    public UnboundedMqttSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public UnboundedReader createReader(PipelineOptions options,
+                                        CheckpointMark checkpointMark) {
+      return new UnboundedMqttReader(this);
+    }
+
+    @Override
+    public List<UnboundedMqttSource> generateInitialSplits(int desiredNumSplits,
+                                                           PipelineOptions options) {
+      List<UnboundedMqttSource> sources = new ArrayList<>();
+      for (int i = 0; i < desiredNumSplits; i++) {
+        // NB: it's important that user understand the impact of MQTT message QoS
+        sources.add(new UnboundedMqttSource(spec));
+      }
+      return sources;
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return VoidCoder.of();
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(byte[].class);
+    }
+
```

Comment:
Clean up blank lines like these too. Generally they are a good idea between members, but not before the first member or after last.